medcat/relation_extraction.py

```diff
@@ -0,0 +1,293 @@
+import json
+import logging
+import os
+import numpy
+import logging
+from numpy.core.fromnumeric import take
+import torch
+
+import torch.nn
+import pickle
+import dill
+import torch.optim
+import torch
+from torch.utils.data import dataloader
+import tqdm
+import torch.nn as nn
+from torch import Tensor
+from datetime import date, datetime
+from torch.nn.modules.module import T
+from transformers import BertConfig
+from ast import literal_eval
+from itertools import permutations
+from pandas.core.series import Series
+from medcat.cdb import CDB
+from medcat.config_re import ConfigRE
+from medcat.utils.relation_extraction.tokenizer import TokenizerWrapperBERT
+
+from spacy.tokens import Doc
+from typing import Dict, Iterable, List, Optional, Set, Tuple, Union
+from transformers import AutoTokenizer
+from torch.utils.data import Dataset, DataLoader
+from medcat.utils.meta_cat.ml_utils import split_list_train_test
+
+
+from medcat.utils.relation_extraction.eval import Two_Headed_Loss
+from medcat.utils.relation_extraction.models import BertModel_RelationExtraction
+from medcat.utils.relation_extraction.pad_seq import Pad_Sequence
+
+from medcat.utils.relation_extraction.utils import batch_split, create_tokenizer_pretrain, load_bin_file, load_results, load_state, put_blanks, save_bin_file, save_results, tokenize
+
+from medcat.utils.relation_extraction.rel_dataset import RelData
+from seqeval.metrics import precision_score, recall_score, f1_score
+
+class RelationExtraction(object):

```

Comment:
To Zeljko's prior comment. This module should be similar to MetaCAT so, i.e. subclasses PipeRunner, so that pipe and __call__ etc. are available. We want ultimately want API like:

CAT(cdb, meta_cats=[ MetaCAT .. ] , rel_cats=[ RelCAT(... ), ... ])